File: 2793-count-the-number-of-complete-components/count-the-number-of-complete-components.py
from typing import List
#edges = (k*(k - 1))/2 where k is nodes
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]

        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        visited = [False] * n
        answer = 0

        def dfs(node):
            visited[node] = True
            nodes = 1
            edge_count = len(graph[node])

            for neighbor in graph[node]:
                if not visited[neighbor]:
                    n_nodes, n_edges = dfs(neighbor)
                    nodes += n_nodes
                    edge_count += n_edges

            return nodes, edge_count

        for i in range(n):
            if not visited[i]:
                nodes, edge_count = dfs(i)

                # Each edge is counted twice
                edge_count //= 2

                if edge_count == nodes * (nodes - 1) // 2:
                    answer += 1

        return answer

File: 2793-count-the-number-of-complete-components/count-the-number-of-complete-components.java
import java.util.*;

class Solution {
    public int countCompleteComponents(int n, int[][] edges) {
        List<Integer>[] graph = new ArrayList[n];

        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int[] edge : edges) {
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0]);
        }

        boolean[] visited = new boolean[n];
        int answer = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                int[] result = dfs(i, graph, visited);

                int nodes = result[0];
                int edgesCount = result[1] / 2; // Each edge counted twice

                if (edgesCount == nodes * (nodes - 1) / 2) {
                    answer++;
                }
            }
        }

        return answer;
    }

    private int[] dfs(int node, List<Integer>[] graph, boolean[] visited) {
        visited[node] = true;

        int nodes = 1;
        int edges = graph[node].size();

        for (int neighbor : graph[node]) {
            if (!visited[neighbor]) {
                int[] result = dfs(neighbor, graph, visited);
                nodes += result[0];
                edges += result[1];
            }
        }

        return new int[]{nodes, edges};
    }
}